README.md

```diff
@@ -120,9 +138,127 @@ module.exports = function (fastify, options, next) {
+2. Update the HMAC object with the signature content string obtained from `getMessage`
+   1. The `getMessage` function returns a formatted string according to [IETF draft standards][1] containing the Signature content listed under `headers` in the `request` Signature header.
+3. Encode the signature with the digest encoding string returned from [`getSignatureEncoding`](#getSignatureEncoding-Method) in the plugin `options` object.
+4. Return the calculated signature string of: `"pQul5YFrqv76Zq2bE1kWjJfFGnTu0MwU7X7c8MWDswAI5V7dROqKbBWKUGcoysxujgTqkJo/Eg74x34o54hqRg=="`
+
+## getDigest Method
+
+The `getDigest` method returns a Digest header string to be used when assembling the HMAC signature input. 
+
+This method is called with two parameters, the fastify `request` object and the plugin `options` object. 
+
+Given the [Example Request](#example-request) this method will:
+1. Parse the `request` Digest header to determine the appropriate hashing algorithm
+2. Hash the `request` body using the algorithm
+3. Apply the digest encoding found in `options` property `digestEncoding` - default: `'base64'`
+3. Compare the new digest value with the value in the `request` Digest header
+   - This will throw an error if the two values do not match
+4. Reconstruct the digest header and return: `"sha-512=+PtokCNHosgo04ww4cNhd4yJxhMjLzWjDAKtKwQZDT4Ef9v/PrS/+BQLX4IX5dZkUMK/tQo7Uyc68RkhNyCZVg=="`
+
+**Note:** The default method assumes:
+1. The Digest value is a hash of only the request body
+2. The body content is always JSON
+3. The request Digest header only contains a single digest value
+
+## getAlgorithm Method
+
+The `getAlgorithm` method returns a HMAC algorithm string to be used when generating a HMAC signature.
+
+This method is called with two parameters, the fastify `request` object and the plugin `options` object.
+
+Given the [Example Request](#example-request) this method will:
+1. Parse the `keyId` and `algorithm` Signature header properties from the `request`
+2. Use the `algorithm` and `keyId` values to look up an algorithm string in `algorithmMap` from the plugin `options`
+3. Return the algorithm string of `"sha512"`
+
+For the default `getAlgorithm`, an `algorithmMap` object matching the following format is expected:
+
+```js
+{
+// const algorithmMap = {
+//   [algorithm]: {
+//     [keyId]: [algorithmString]
+//   } 
+// }
+
+const algorithmMap = {
+  hs2019: {
+    'test-key-a': 'sha512',
+    'test-key-b': 'sha256'
+  } 
+}
+```
+
+## getSignatureEncoding Method
+
+The `getSignatureEncoding` method returns a digest encoding string to be used when generating a HMAC signature.
+
+This method is called with two parameters, the fastify `request` object and the plugin `options` object.
+
+The default method simply returns `"base64"`. 
+
+# Replacing Default Methods
+
+## Example: Shopify HMAC Query Parameter Verification
```

Comment:
Should this be in the examples director in a readme? We could just link to there for additional examples. I'd like to keep the main readme lean if possible. Thoughts?

Reply:
I don't mind having the Shopify example top-level but I would pare it down a little & link out to the Shopify docs about how it's constructed